kong/db/dao/certificates.lua

```diff
@@ -0,0 +1,209 @@
+local singletons = require "kong.singletons"
+local cjson      = require "cjson"
+local utils      = require "kong.tools.utils"
+
+-- Get an array of server names from either a string (split+sort),
+-- an array(sort) or ngx.null(return {})
+-- Returns an error if the list has duplicates
+-- Returns nil if input is falsy.
+local function parse_name_list(input, errors)
+  local name_list
+  if type(input) == "string" then
+    name_list = utils.split(input, ",")
+  elseif type(input) == "table" then
+    name_list = utils.shallow_copy(input)
+  elseif input == ngx.null then
+    name_list = {}
+  end
+
+  if not name_list then
+    return nil
+  end
+
+  local found = {}
+  for _, name in ipairs(name_list) do
+    if found[name] then
+      local msg   = "duplicate server name in request: " .. name
+      local err_t = errors:invalid_input(msg)
+      return nil, tostring(err_t), err_t
+    end
+    found[name] = true
+  end
+
+  table.sort(name_list)
+  return setmetatable(name_list, cjson.empty_array_mt)
+end
+
+
+local _Certificates = {}
+
+-- Creates a certificate
+-- If the provided cert has a field called "server_names" it will be used to generate server
+-- names associated to the cert, after being parsed by parse_name_list.
+-- Returns a certificate with the server_names sorted alphabetically.
+function _Certificates:insert_with_name_list(cert)
+  local db = singletons.db
+  local name_list, err, err_t = parse_name_list(cert.server_names, self.errors)
+  if err then
+    return nil, err, err_t
+  end
+
+  if name_list then
+    local ok, err, err_t = db.server_names:check_list_is_new(name_list)
+    if not ok then
+      return nil, err, err_t
+    end
+  end
+
+  cert.server_names = nil
+  cert, err, err_t = assert(self:insert(cert))
+  if not cert then
+    return nil, err, err_t
+  end
+  cert.server_names = name_list or cjson.empty_array
+
+  if name_list then
+    local ok, err, err_t = db.server_names:insert_list({id = cert.id}, name_list)
+    if not ok then
+      return nil, err, err_t
```

Comment:
This leads to interesting situations, like certificate added, but no server_names added. Do we have any concerns of that?

Reply:
I would rather not have that situation, but I honestly didn't know what else to do (we don't have transactions at this level, because Cassandra).

This code reflects what the old DAO code did - it makes as much "checks" as possible before writing anything to the db.